cmd/geoipupdate/main.go

```diff
@@ -0,0 +1,428 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	flock "github.com/theckman/go-flock"
+)
+
+// Version is the program's version number.
+const Version = "0.0.1"
+
+func main() {
+	log.SetFlags(0)
+
+	args := getArgs()
+
+	config, err := NewConfig(args.ConfigFile, args.DatabaseDirectory)
+	if err != nil {
+		log.Fatalf("Error loading configuration file: %+v", err)
+	}
+
+	lock, err := setup(config, args.Verbose)
+	if err != nil {
+		log.Fatalf("Error preparing to update: %+v", err)
+	}
+	defer func() {
+		if err := lock.Unlock(); err != nil {
+			log.Fatalf("Error unlocking lock file: %+v", errors.Wrap(err, "unlocking"))
+		}
+	}()
+
+	if err := run(config, args.Verbose); err != nil {
+		log.Fatalf("Error retrieving updates: %+v", err)
+	}
+}
+
+func setup(
+	config *Config,
+	verbose bool,
+) (*flock.Flock, error) {
+	if err := maybeSetProxy(config); err != nil {
+		return nil, err
+	}
+
+	if err := checkEnvironment(config); err != nil {
+		return nil, err
+	}
+
+	lock := flock.NewFlock(config.LockFile)
+	ok, err := lock.TryLock()
+	if err != nil {
+		return nil, errors.Wrap(err, "error acquiring a lock")
+	}
+	if !ok {
+		return nil, errors.Errorf("could not acquire lock on %s", config.LockFile)
+	}
+	if verbose {
+		log.Printf("Acquired lock file lock (%s)", config.LockFile)
+	}
+
+	return lock, nil
+}
+
+var client = &http.Client{
+	Timeout: 30 * time.Minute, // Allow for slow downloads
+}
+
+func maybeSetProxy(config *Config) error {
+	if config.Proxy == "" {
+		return nil
+	}
+	proxy := config.Proxy
+
+	if strings.Index(proxy, ":") == -1 {
+		proxy += ":1080" // The 1080 default came from cURL.
+	}
+	proxy = "http://" + proxy
```

Comment:
I wonder if we should allow people to pass in a full proxy URL. It looks like Go also supports `socks5` as the scheme. That could be a later improvement though.

Reply:
Yeah, we could do that. I expected it would be pretty rare as the only examples we have show without a protocol. It's easy enough to support socks5 though, so let's do it!